src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs

```diff
@@ -0,0 +1,285 @@
+    /// <remarks>
+    /// We don't use 'NonBacktracking' here because it is not supported by the
+    /// regex source generator. However these high-performance patterns should
+    /// not cause problematic backtracking. They are extremely simple, anchored
+    /// to the front with at most one optional group at the end. Generated code
+    /// has been inspected to see that the only "loops" have at most 1
+    /// iteration. If more complex regexes are added, the generated code should
+    /// be re-examined.
+    ///
+    /// 'SingleLine' is used to ensure that '.' matches any character including
+    /// new lines. This is used to share regexes that can purely skip over the
+    /// signature that has already been matched before the regex is applied.
+    ///
+    /// Also note that the high-performance scanner does not allow substitution
+    /// of the regex engine and the regex engine built-in to .NET is always
+    /// used. Profiling showed no performance gain from RE2, even on .NET
+    /// Framework, even for the unoptimized 'SecretMasker' using only the
+    /// highly-identifiable patterns, but this merits more scrutiny and
+    /// investigation.
+    ///
+    /// Even more critically, the regex engine abstraction does not currently
+    /// provide a mechanism to search within a substring without allocation.
+    /// There is no incentive to add it, which would require work in the RE2
+    /// wrapper, unless further investigation reveals a flaw in the finding on
+    /// RE2 performance above.
+    /// </remarks>
+    private const RegexOptions Options = RegexOptions.Compiled |
+                                         RegexOptions.ExplicitCapture |
+                                         RegexOptions.CultureInvariant |
+                                         RegexOptions.Singleline;
+
+
+    public CompiledHighPerformancePattern(string signature,
+                                           int signaturePrefixLength,
+                                           int minMatchLength,
+                                           int maxMatchLength,
+                                           Regex scopedRegex)
+    {
+#if HIGH_PERFORMANCE_CODEGEN // This validation happens at compile time and needn't be repeated at runtime.
+        if (!scopedRegex.ToString().StartsWith("^"))
+        {
+            throw new ArgumentException($"The regular expression must be anchored to the beginning of the input: '{scopedRegex}'", nameof(scopedRegex));
+        }
+
+        if (signature.Length != 3 && signature.Length != 4)
+        {
+            throw new ArgumentException($"The signature must be 3 or 4 characters long: '{signature}'.", nameof(signature));
+        }
+#endif
+        char s0 = signature[0];
+        char s1 = signature[1];
+        char s2 = signature[2];
+        char s3 = signature.Length < 4 ? '\0' : signature[3];
+
+#if HIGH_PERFORMANCE_CODEGEN
+        if (s0 > 0x7F || s1 > 0x7F || s2 > 0x7F || s3 > 0x7F)
```

Comment:
>s0 > 0x7F || s1 > 0x7F || s2 > 0x7F || s3 > 0x7F [](http://example.com/codeflow?start=12&length=48)

seems like someone somewhere someday might also want this helper, as some inlinable utility